Implement a gpgme-based crypto plugin allowing to use PGP keys

dcaliste opened the pull request:
Hello @Venemo, @chriadam and @jpetrell. This is a pull request to show the current state of affair on the GnuPG plugin. It's not ready yet and I'm still working on it. It currently contains basics to generate a key and sign with it (not fully working yet though). Later on I'll add the code to create a pinentry as discussed in #5.

It can serve as base for the discussion of tomorrow's meeting though if you are still fit after the sprint of recent days ;)

Congrats for the work by the way.

chriadam commented on tests/plugins/testgnupgplugin/tst_gnupgplugin.cpp:
```diff
@@ -0,0 +1,225 @@
+    QSignalSpy reqstat(&req, &StoredKeyIdentifiersRequest::statusChanged);
+    QSignalSpy reqiden(&req, &StoredKeyIdentifiersRequest::identifiersChanged);
+
+    // req.setCryptoPluginName(OPENPGP_PLUGIN);
```
With the refactor, you need to do: setStoragePluginName(OPENPGP_PLUGIN); and also that plugin must also implement the EncryptedStoragePlugin interface (as well as the CryptoPlugin interface) -- see e.g. the SqlCipher plugin for example.

dcaliste replied:
Thanks, I've added implementation for the EncryptedStorage API.

dcaliste commented:
@chriadam I've updated my PR. It's now creating PGP keys, listing, signing and verifying.

I've some remarks on the API that we can discuss tomorrow.

dcaliste commented:
For reference, this is how the pinentry dialog looks like when doing a PGP signing operation from command line with ```gpg2``` executable:
![20180522120306](https://user-images.githubusercontent.com/5889532/40356215-b858b458-5db8-11e8-8b98-67cde140db92.png)

Venemo commented:
Congrats @dcaliste that does look very cool. Thank you for doing this!

dcaliste commented:
And now, also from the command-line tool:
<pre>
[nemo@Sailfish devel]$ secrets-tool --sign org.sailfishos.crypto.plugin.gnupg.openpgp org.sailfishos.crypto.plugin.gnupg.openpgp "Damien Caliste <dcaliste@free.fr>" F58016BD85212EFF4763BC70E7F32BC76D8C1C73 SHA256 toto
-----BEGIN PGP SIGNATURE-----
Version: GnuPG v2.0.4 (GNU/Linux)

iQGVAwUBWwaoqefzK8dtjBxzAQJAvQwAsj5Gynbp4nLbYuKuN2wCQNGKRuxGjP8V
SnqWKzJ9UI9qbZrKmX86wx5mbdT7vjODHqjpu0Ua05/RVn95Gy9F8GrBjDuuEX+N
OflUnf8juwFcgig3m3lcbgyQFcqt9eGFbm41lKZuLh+MD5Q+opRcLA0wG6OLYpr9
30xuOvd74x8wlvrymxP6gslZoOnCelT2lZsCuMdTE01tPvNsVKqq0gduKIBYYWv5
giwO5GMge29f+lwr9BgwZGqcVHrL5Lc7VHITPHtGKqIYzZpA93gDn/sNQ07/F98V
XLuR3ze2z7F3InbChemRdIaQm7bWUiSmh5MdIW0x08ESezWqdlv1FxlcHOOc3dXh
0l2Us9kyTHEH5S4qIhXaY+UBvF544uQgUFYNPOW0JFsIaK72HrKNeNE0fNVuqYHX
dLDEGH5CxcsP2KZvoFtx26kLkbgLZJOsYvE5aIynkCTCaNO1fKcK3W1J1nBLQ5qd
Skz+aXwU1wRPNc4zw2TYMwUwVV/L+2Fr
=TqYy
-----END PGP SIGNATURE-----
</pre>

One can also list the available keys:
<pre>
[nemo@Sailfish devel]$ secrets-tool --list-keys org.sailfishos.crypto.plugin.gnupg.openpgp
--------------------------------------------
Storage plugin: "org.sailfishos.crypto.plugin.gnupg.openpgp"
    Collection: "Mattia Rizzolo <mattia@mapreri.org>"
          Name: "66AE2B4AFCCF3F52DA184D184B043FCDB9444540"
--------------------------------------------
Storage plugin: "org.sailfishos.crypto.plugin.gnupg.openpgp"
    Collection: "Mattia Rizzolo <mattia@mapreri.org>"
          Name: "6E9BB401D6EC87D3FD1F4EAD39446A919948BA4E"
--------------------------------------------
Storage plugin: "org.sailfishos.crypto.plugin.gnupg.openpgp"
    Collection: "Mattia Rizzolo <mattia@mapreri.org>"
          Name: "8B7868786C33E5C64C4D0A480816B9E18C762BAD"
--------------------------------------------
Storage plugin: "org.sailfishos.crypto.plugin.gnupg.openpgp"
    Collection: "Mattia Rizzolo <mattia@mapreri.org>"
          Name: "4A51F45E4DC34E312FC79C9ABC3F58C272FC5DA4"
--------------------------------------------
Storage plugin: "org.sailfishos.crypto.plugin.gnupg.openpgp"
    Collection: "Damien Caliste <dcaliste@free.fr>"
          Name: "F58016BD85212EFF4763BC70E7F32BC76D8C1C73"
</pre>

monich commented on plugins/gnupgplugin/gpgme_p.h:
```diff
@@ -0,0 +1,273 @@
+                        qWarning() << "protocol:" << gpgme_get_protocol_name(info->protocol);
+                        if (info->file_name && !info->version)
+                            qWarning() << "engine " << info->file_name
+                                       << " not installed properly";
```
I believe `QDebug` is inserting spaces between the strings, there's no need to add more.

Venemo replied:
Yep.

dcaliste commented:
And now, also allowing PGP encryption / decryption:
<pre>
[nemo@Sailfish devel]$ cat data0
Test plaintext data
[nemo@Sailfish devel]$ secrets-tool --encrypt org.sailfishos.crypto.plugin.gnupg.openpgp org.sailfishos.crypto.plugin.gnupg.openpgp "Damien Caliste <dcaliste@free.fr>" F58016BD85212EFF4763BC70E7F32BC76D8C1C73 data0 | tee data0.enc
IV:
LS0tLS1CRUdJTiBQR1AgTUVTU0FHRS0tLS0tClZlcnM=
aW9uOiBHbnVQRyB2Mi4wLjQgKEdOVS9MaW51eCkKCmg=
UUdNQXl0RTlacXJtN1N6QVF2L1VyVWxrNXlXSjMrK2g=
VWFCYnZvTmVkcVNwM3hyVUxRdmtIWjZFdHhMSnFxeAo=
[…]
[nemo@Sailfish devel]$ secrets-tool --decrypt org.sailfishos.crypto.plugin.gnupg.openpgp org.sailfishos.crypto.plugin.gnupg.openpgp "Damien Caliste <dcaliste@free.fr>" F58016BD85212EFF4763BC70E7F32BC76D8C1C73 data0.enc
Test plaintext data

chriadam commented on plugins/gnupgplugin/gpgme_p.h:
```diff
@@ -0,0 +1,277 @@
+ * Contact: Damien Caliste <dcaliste@free.fr>
+ * All rights reserved.
+ * BSD 3-Clause License, see LICENSE.
+ */
```
best to define a header guard even for private headers with known inclusion locations.

chriadam commented on plugins/gnupgplugin/gpgme_p.h:
```diff
@@ -0,0 +1,277 @@
+            if (!err2) {
+                while (info) {
+                    if (info->protocol == protocol) {
+                        qWarning() << "protocol:" << gpgme_get_protocol_name(info->protocol);
```
instead of qWarning, best to use qCWarning() and define a plugin-specific logging category.  Hrm, or perhaps we need to add a generic secrets daemon plugin logging category, which all plugins should use...

dcaliste replied:
I've added a commit adding logging category for plugins and used it accordingly.

chriadam commented on plugins/gnupgplugin/gpgme_p.h:
```diff
@@ -0,0 +1,277 @@
+    {
+        return ctx;
+    }
+    QString error()
```
can this be const or is either gpgme_error_code(err) or gpgme_strerror(err) nonconst?

chriadam commented on plugins/gnupgplugin/gpgme_p.h:
```diff
@@ -0,0 +1,277 @@
+        : data(0), err(0)
+    {
+        err = gpgme_data_new_from_mem(&data, origin.constData(), origin.length(), 0);
+        if (gpgme_err_code(err) != GPG_ERR_NO_ERROR)
```
Sailfish OS coding style recommends using braces even for single-line conditional expression blocks.  Also elsewhere in this file / PR.

dcaliste replied:
I've tried to change this everywhere. Hopefully I've not missed a place.

chriadam commented on plugins/gnupgplugin/gpgme_p.h:
```diff
@@ -0,0 +1,277 @@
+        }
+        return false;
+    }
+    const char* fpr() const
```
what does fpr stand for?

dcaliste replied:
fingerprint, changed everywhere.

chriadam commented on plugins/gnupgplugin/gpgmebase.cpp:
```diff
@@ -0,0 +1,761 @@
+        return Result(Result::CryptoPluginKeyGenerationError, ctx.error());
+
+    QString gnupgKeyParms = "<GnupgKeyParms format=\"internal\">\n";
+    if (kpgParams.keyPairType() == KeyPairGenerationParameters::KeyPairDsa) {
```
in this case, you could do: DsaKeyPairGenerationParameters dkpgp(kpgParams); and then read the key-length from the modulusLength() property of the dkpgp instance.

dcaliste replied:
In fact, for DSA, GnuPG handle only length of 1024. I've added a warning in case the user would have chosen another value.

chriadam commented on plugins/gnupgplugin/gpgmebase.cpp:
```diff
@@ -0,0 +1,761 @@
+        gnupgKeyParms += "Subkey-Length: 1024\n";
+    } else if (kpgParams.keyPairType() == KeyPairGenerationParameters::KeyPairRsa) {
+        gnupgKeyParms += "Key-Type: RSA\n";
+        gnupgKeyParms += QStringLiteral("Key-Length: %1\n").arg(static_cast<RsaKeyPairGenerationParameters>(kpgParams).modulusLength());
```
I'm surprised the static cast works, actually... maybe better to simply in-place construct: RsaKeyPairGenerationParameters(kpgParams).modulusLength();

dcaliste replied:
Updated.

chriadam commented on plugins/gnupgplugin/gpgmebase.cpp:
```diff
@@ -0,0 +1,761 @@
+ 
+        err = gpgme_op_genkey(ctx, gnupgKeyParms.toUtf8().constData(), pub, priv);
+
+        // Todo do something with the public data.
```
is the public data not already used above with the gpgme_op_genkey() call?

dcaliste replied:
I've updated the comment. Here it's to do something with the generated data from gpgme that is a certificate request in S/MIME.

chriadam commented on plugins/gnupgplugin/gpgmebase.cpp:
```diff
@@ -0,0 +1,761 @@
+                  ? GPGmeKey::Secret : GPGmeKey::Public);
+    if (!gkey) {
+        // Todo: need to set the storage error code to
+        // Sailfish::Secrets::Result::InvalidSecretError
```
there should be a constructor for this now :-)

dcaliste replied:
I forgot to upgrade this ;)

chriadam commented on plugins/gnupgplugin/gpgmebase.cpp:
```diff
@@ -0,0 +1,761 @@
+        return Result(Result::CryptoPluginDecryptionError,
+                      QStringLiteral("cannot decrypt message: %1.").arg(gpgme_strerror(err)));
+    }
+    Result verifyResult = verify(CryptoManager::OperationDecrypt,
```
What does this do?

dcaliste replied:
Name of function changed for checkGPGmeVerification() which is hopefully more descriptive.

dcaliste commented on rpm/sailfish-secrets.spec:
```diff
@@ -225,6 +225,9 @@ BuildRequires:  pkgconfig(Qt5DBus)
 BuildRequires:  pkgconfig(dbus-1)
 BuildRequires:  pkgconfig(libcrypto)
 BuildRequires:  qt5-plugin-sqldriver-sqlite
+BuildRequires:  pkgconfig(libgpg-error)
```
That's wise. Changed for this.

chriadam replied:
After more thought, is there any way to implement the pinentry component such that it can be licensed with BSD-3-clause also?  Given that we were hoping to distribute this gnupg plugin along with the daemon, I'd prefer a maximally permissive license to avoid potential inadvertent violations...

dcaliste replied:
I understand, but the pinentry program is not linking with the daemon. In fact, it ideally should be part of GnuPG itself which is GPL, like other pinentries. Indeed, the pinentry program is used by GnuPG executable to ask for user entry and is using IPC (via libassuan which is LGPL) to interface with GnuPG. So it may even be whatever license.

Our pinentry implementation is then linking with libsecret to interface nicely with SFOS environment. But since libsecret is BSD licensed, the pinentry can be whatever restrictive license we want.

So I think putting the pinentry under GPL won't violate anything. If you think differently, I can change it for BSD though.

monich commented on plugins/gnupgplugin/gpgme_p.h:
```diff
@@ -0,0 +1,277 @@
+            return;
+        }
+        if (home.isEmpty())
+            return;
```
Return just like that - without releasing the context or anything? Or should this check be done earlier, before the context is allocated?

dcaliste replied:
The context will be automatically freed when the structure will become out of scope by the structure destructor.

dcaliste commented:
I've amended the commits but I still didn't have time to check that it's working and test. Don't approve the PR immediately. I'll do it tomorrow. Thanks for the review.

dcaliste commented on plugins/gnupgplugin/gpgme_p.h:
```diff
@@ -0,0 +1,277 @@
+    GPGmeData(const QByteArray &origin)
+        : data(0), err(0)
+    {
+        err = gpgme_data_new_from_mem(&data, origin.constData(), origin.length(), 0);
```
In fact, no copy is done, see the 0 argument on last position. Doc in https://www.gnupg.org/documentation/manuals/gpgme/Memory-Based-Data-Buffers.html

I may #define NO_COPY 0 to be clear...

monich replied:
Then if it's not copy, then you need to carry QByteArray around too, to make sure that the actual data don't get deallocated before GPGmeData.

dcaliste replied:
These are just helpers in gpgme_p.h to automatically have deallocation. I kept the same restrictions than in gpgme API. Callers are responsible to keep the data alive indeed. Since these helpers are not passed between routines I think it's fine.

dcaliste replied:
I've added a comment to relmind that origin QByteArray object should be alive for the whole life of the created structure and replaced the last 0 argument with a NO_COPY #define.

Thank you for your comments.

dcaliste commented:
@chriadam ok, I've quickly tested on device that signing and verifying are working with the secrets-tool CLI, new automatic test for GnuPG is also passing on device. We can proceed.

dcaliste commented:
Added encryption / decryption tests and rebased.

dcaliste commented:
I've modified a bit the pinentry code to handle caching and added also a DConf entry to set wether or not the user wants to store passphrases.

I've changed the ```fprintf``` for qCDebug() also.

dcaliste commented:
@chriadam I've discovered a funny issue: when running daemon from CLI (with a simple `devel-su -p sailfish-secretsd`), the GnuPG plugins were properly loaded. But when launched by systemd, no GnuPG plugins.

It comes from the fact that invoker is currently running the daemon as `type=qt5`, while GnuPG plugins are using non-qt5 libraries (like gpgme). So I've changed the service file to run invoker with `type=generic`, I hope it's not an issue.

chriadam commented:
Regarding changing it to a generic booster, that's perfectly fine.  Aside from that, also LGTM, although I'm concerned that I don't know about about the implications of a GPLv2 licensed plugin given that we intend to distribute the plugin along with the daemon (or is it ok, given that the daemon is even more permissively licensed?  IANAL...)

dcaliste commented:
Thank you @chriadam very much for looking at this.

GPGme has two licenses as far as I can see from the two COPYING and COPYING.LESSER in the sources. One of them is LGPL, so there is no issue linking a plugin of a BSD executable with a LGPL library (read GPGme). I think that the plug-in is written with BSD headers, right ?

The pinentry program is another matter since it is a program, interfacing with GnuPG via IPC and linking the BSD sailfish-secret library. So this program can be whatever we want without any issues. I've chosen the GPL, but I can change my mind if there are reasons that I didn't foresee.

chriadam commented:
I'd prefer that each package used just a single license, and pinentry is packaged with the plugin in this case; thus, if possible, I would prefer it to be licensed as BSD also.  Basically, I have an aversion to accepting any code into the repository which has a less permissive license than that of the rest, mostly because I don't entirely understand the ramifications of those different licenses, and what they might mean e.g. for our partners also.

Finally, it simplifies communication with partner vendors; otherwise, if there is a package which has multiple licenses, someone would have to potentially analyse each source file involved in generating the package to determine which parts may or may not be involved, and what the potential implications might be, when making a decision about what risks are implied when generating a device image with the software included.

Of course, it's your contribution, so if you strong prefer to keep pinentry as GPL in this case, I will simply ask Jolla Legal to definitively look into this case to give the green light, and then once that happens I will merge as-is.

chriadam commented:
A couple of things I noticed after rebasing on top of master and running the tests:
1) you should generate a test plugin (i.e. whose name ends with ".test") similarly to the other plugins, so that we can run the unit test after starting the daemon in --test mode
2) on my device there is a failure in the addSubKey() test:

QDEBUG : tst_gnupgplugin::addSubKey() [D] unknown:0 - "Collection John Doe <john.doe@example.com> does not exist"
FAIL!  : tst_gnupgplugin::addSubKey() Compared values are not the same
   Actual   (dskr.result().code()): Failed
   Expected (Result::Succeeded)   : Succeeded
   Loc: [tst_gnupgplugin.cpp(193)]

Do I need to install some package which I'm missing?

dcaliste commented:
> on my device there is a failure in the addSubKey() test:

That's expected. It's due to the fact that DeleteStoredKeyRequest is not passing the custom parameters (because DeleteSecretRequest from Secrets don't have it, only requests from Crypto, and the custom parameters are dropped when calling the DeleteSecretRequest from the DeleteStoredKeyRequest). We discussed this a long time ago. I proposed to expand the Secrets plugin API, but I didn't find the time to actually propose a PR for it.

So for the moment, I've disabled this part of the test and put a comment about the API limitation.

dcaliste commented:
> you should generate a test plugin (i.e. whose name ends with ".test")

Ok, done for the OpenPGP plugin. What can I run to check if the plugin passes the unit tests ? The daemon with `--test` is properly able to load the test plugin but I don't know what to run after. Besides, I'm guessing it will not work out of the box because key generation for OpenPGP is a bit specific and requires some custom parameters besides key name and collection name.

chriadam commented:
Ok, sounds good.  When running the daemon in --test mode, I just expect to be able to run the tst_gnupgplugin test and have it pass.  If that now works, then LGTM.  (I'll double check that on my device this afternoon.)  The other unit tests should still pass (i.e. they shouldn't fail just because this package is installed) but they needn't actually test any functionality from this plugin.

chriadam commented:
Thank you very much.  Merging now.
